Architecture/TodoApp/android/core/src/main/java/com/novoda/todoapp/tasks/service/PersistedTasksService.java

```diff
@@ -170,45 +175,67 @@ public Tasks call(List<Task> tasks) {
     }
 
     @Override
-    public Action0 refreshTasks() {
-        return new Action0() {
+    public void refreshTasks() {
+        fetchRemote()
+                .compose(asValidatedEvent(taskRelay.getValue()))
+                .subscribeOn(scheduler)
+                .subscribe(taskRelay);
+    }
+
+    private Observable<Tasks> fetchRemote() {
+        return Observable.defer(new Func0<Observable<Tasks>>() {
             @Override
-            public void call() {
-                fetchFromRemote()
-                        .compose(asValidatedEvent(taskRelay.getValue()))
-                        .subscribe(taskRelay);
+            public Observable<Tasks> call() {
+                return fetchFromRemote(clock.timeInMillis());
             }
-        };
+        });
     }
 
     @Override
-    public Action0 clearCompletedTasks() {
-        return new Action0() {
+    public void clearCompletedTasks() {
+        long actionTimestamp = clock.timeInMillis();
+
+        remoteDataSource.clearCompletedTasks()
+                .compose(confirmLocalDeletionsOrRevert(taskRelay, actionTimestamp))
+                .map(updateOnlyMostRecent(actionTimestamp))
+                .compose(asValidatedEvent(taskRelay.getValue()))
+                .doOnNext(taskRelay)
+                .flatMap(persistEventTasksLocally())
+                .subscribeOn(scheduler)
+                .subscribe();
+    }
+
+    private Func1<Tasks, Tasks> updateOnlyMostRecent(final long actionTimestamp) {
+        return new Func1<Tasks, Tasks>() {
             @Override
-            public void call() {
-                long actionTimestamp = clock.timeInMillis();
-                Event<Tasks> currentState = taskRelay.getValue();
-                Tasks allLocalTasks = currentState.data().or(Tasks.empty());
-                Tasks uncompletedLocalTasks = mapFunctionToTasks(allLocalTasks, markCompletedAsDeleted());
+            public Tasks call(Tasks tasks) {
+                Tasks currentTasks = taskRelay.getValue().data().or(Tasks.empty());
+                return tasks.insertOrUpdate(currentTasks.filter(isMoreRecentAction(tasks, actionTimestamp)));
```

Comment:
I would really extract this to a `mostRecentTasks` variable and then use it to update `tasks`.